crates/wysiwyg/src/tests/test_to_markdown.rs

```diff
@@ -0,0 +1,228 @@
+    // Internal spaces are preserved.
+    assert_to_md("abc   def", "abc   def");
+}
+
+#[test]
+fn text_with_linebreaks() {
+    // One new line.
+    assert_to_md(
+        "abc<br />def",
+        r#"abc\
+def"#,
+    );
+
+    // Two new lines (isn't transformed into a new block).
+    assert_to_md(
+        "abc<br /><br />def",
+        r#"abc\
+\
+def"#,
+    );
+}
+
+#[test]
+fn text_with_italic() {
+    assert_to_md("<em>abc</em>", "*abc*");
+    // Internal emphasis.
+    assert_to_md("abc<em>def</em>ghi", "abc*def*ghi");
+    assert_to_md("abc <em>def</em> ghi", "abc *def* ghi");
+    assert_to_md_no_roundtrip("abc<em> def </em>ghi", "abc* def *ghi");
+    assert_to_md(
+        "abc <em>line1<br />line2<br /><br />line3</em> def",
+        r#"abc *line1\
+line2\
+\
+line3* def"#,
+    );
+}
+
+#[test]
+fn text_with_bold() {
+    assert_to_md("<strong>abc</strong>", "__abc__");
+    assert_to_md_no_roundtrip("abc<strong>def</strong>ghi", "abc__def__ghi");
+    assert_to_md("abc <strong>def</strong> ghi", "abc __def__ ghi");
+    assert_to_md_no_roundtrip(
+        "abc<strong> def </strong>ghi",
+        "abc__ def __ghi",
+    );
+    assert_to_md(
+        "abc <strong>line1<br />line2<br /><br />line3</strong> def",
+        r#"abc __line1\
+line2\
+\
+line3__ def"#,
+    );
+}
+
+#[test]
+fn text_with_italic_and_bold() {
+    assert_to_md("<em><strong>abc</strong></em>", "*__abc__*");
+    assert_to_md("<em>abc <strong>def</strong></em> ghi", "*abc __def__* ghi");
+    assert_to_md(
+        "abc <em><strong>line1<br />line2</strong> def</em>",
+        r#"abc *__line1\
+line2__ def*"#,
+    );
+}
+
+#[test]
+fn text_with_strikethrough() {
+    assert_to_md("<del>abc</del>", "~~abc~~");
+    assert_to_md_no_roundtrip("abc<del>def</del>ghi", "abc~~def~~ghi");
+    assert_to_md("abc <del>def</del> ghi", "abc ~~def~~ ghi");
+    assert_to_md_no_roundtrip("abc<del> def </del>ghi", "abc~~ def ~~ghi");
+    assert_to_md(
+        "abc <del>line1<br />line2<br /><br />line3</del> def",
+        r#"abc ~~line1\
+line2\
+\
+line3~~ def"#,
+    );
+}
+
+#[test]
+fn text_with_underline() {
+    assert_to_md_no_roundtrip("<u>abc</u>", "abc");
+}
+
+#[test]
+fn text_with_inline_code() {
+    assert_to_md("<code>abc</code>", "`` abc ``");
+    // Inline code with a backtick inside.
+    assert_to_md("<code>abc ` def</code>", "`` abc ` def ``");
+    // Inline code with a backtick at the start.
+    assert_to_md("<code>`abc</code>", "`` `abc ``");
+    assert_to_md("abc <code>def</code> ghi", "abc `` def `` ghi");
+    assert_to_md("abc<code> def </code>ghi", "abc``  def  ``ghi");
+    // It's impossible to get a line break inside an inline code with Markdown.
+    assert_to_md_no_roundtrip(
+        "abc <code>line1<br />line2<br /><br />line3</code> def",
+        "abc `` line1 line2  line3 `` def",
+    );
+    assert_to_md_no_roundtrip(
+        "abc <code>def <strong>ghi</strong> jkl</code> mno",
+        "abc `` def __ghi__ jkl `` mno",
```

Comment:
Is this valid markdown? Maybe we should strip formatting inside code blocks?

Reply:
So… yeah, having 2 backticks + 1 space as a surroundding pair is valid! See [the specification](https://github.github.com/gfm/#example-339). Using 2 backticks allows to have one backtick inside the code. It's the most weird-syntax-proof syntax I can come with. Example, to write  `` `foo _bar_ baz` `` I need to write:

```md
`` `foo _bar_ baz` ``
```

But why can't ensure a roundtrip here?

The HTML code is valid. The resulting Markdown is valid. But the specification is ambiguous here.

The following HTML code `<code>foo <em>bar</em> baz</code>` must map to `` `foo _bar_ baz` `` in Markdown. That's correct. However, `` `foo _bar_ baz` `` doesn't necessarily map to `<code>foo <em>bar</em> baz</code>` back to HTML. Actually, let's see what the Github compiler does: “`foo _bar_ baz`”. It renders this `foo \_bar\_ baz` (without the backslash). The `_` isn't compiled to `<em>`. Annoying right? That's because HTML allows inline element inside a `<code>` element, but Markdown doesn't.

Reply:
Yeah, which is fine: we know that markdown does not support all of HTML. So arguably the right behaviour is to render

```
abc <code>def <strong>ghi</strong> jkl</code> mno
```

as

```
abc `` def ghi jkl `` mno
```

but it's a corner case and not something we need to do right now.